### PR TITLE
Remove NativeTokensBalanceDto and BalanceDto

### PR DESCRIPTION
**Pull request by thibault-martinez**

### CODE REVIEW

**Review by qrayven — DISMISSED** [review]

File: sdk/src/wallet/account/types/balance.rs [hunk 1]
```diff
@@ -156,59 +114,34 @@ pub struct NativeTokensBalance {
...
-    pub(crate) metadata: Option<MetadataFeature>,
     /// Total amount
     #[getset(get_copy = "pub")]
     pub(crate) total: U256,
     /// Balance that can currently be spent
     #[getset(get_copy = "pub")]
     pub(crate) available: U256,
+    /// Token foundry immutable metadata
+    #[getset(get = "pub")]
+    pub(crate) metadata: Option<String>,
```
Comment (DaughterOfMars) [inline sdk/src/wallet/account/types/balance.rs]: Wouldn't it be better to keep the concrete type?

Comment (thibault-martinez) [reply sdk/src/wallet/account/types/balance.rs]: Added back, with serde support for option_string

File: sdk/src/types/block/output/feature/metadata.rs [hunk 2]
```diff
@@ -31,6 +31,14 @@ impl TryFrom<Vec<u8>> for MetadataFeature {
     }
 }
 
+impl FromStr for MetadataFeature {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::new(prefix_hex::decode::<Vec<u8>>(s).map_err(Error::Hex)?)
```
Comment (DaughterOfMars) [inline sdk/src/types/block/output/feature/metadata.rs:38]: Could change `new` to take `impl Into<Box<[u8]>>` while you're in here.

File: sdk/src/wallet/account/types/balance.rs [hunk 3]
```diff
@@ -122,7 +122,8 @@ pub struct NativeTokensBalance {
     pub(crate) available: U256,
     /// Token foundry immutable metadata
     #[getset(get = "pub")]
-    pub(crate) metadata: Option<String>,
+    #[serde(with = "crate::utils::serde::option_string")]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/account/types/balance.rs:125]: shouldn't this use the serde impl of `MetadataFeature`?

Comment (DaughterOfMars) [reply sdk/src/wallet/account/types/balance.rs:125]: and `serde(default)`

Comment (thibault-martinez) [reply sdk/src/wallet/account/types/balance.rs:125]: No, that would print a packable BoxedSlicePrefix

File: sdk/src/types/block/output/feature/metadata.rs (hunk 2 above)
Comment (thibault-martinez) [reply sdk/src/types/block/output/feature/metadata.rs:38]: And also change the TryFrom then?

Comment (DaughterOfMars) [reply sdk/src/types/block/output/feature/metadata.rs:38]: I don't know about that

File: sdk/src/wallet/account/types/balance.rs (hunk 3 above)
Comment (DaughterOfMars) [reply sdk/src/wallet/account/types/balance.rs:125]: Any reason `MetadataFeature` can't use a prefix hex serde impl?

File: sdk/src/types/block/output/feature/metadata.rs (hunk 2 above)
Comment (thibault-martinez) [reply sdk/src/types/block/output/feature/metadata.rs:38]: Doing the box breaks some nice usecase like `MetadataFeature::new("Hello, World!")`

File: sdk/src/wallet/account/types/balance.rs (hunk 3 above)
Comment (thibault-martinez) [reply sdk/src/wallet/account/types/balance.rs:125]: It's just that most of the time we use the DTO instead, so we'd do the custom serde impl just for this use case

File: sdk/src/types/block/output/feature/metadata.rs (hunk 2 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/output/feature/metadata.rs:38]: We could just add a `impl From<&str> for MetadataFeature` and call `MetadataFeature::from("Hello, World!")` instead.

File: sdk/src/wallet/account/types/balance.rs (hunk 3 above)
Comment (DaughterOfMars) [reply sdk/src/wallet/account/types/balance.rs:125]: That DTO can't be deleted?

Comment (thibault-martinez) [reply sdk/src/wallet/account/types/balance.rs:125]: No, it serialises the type field as well

Comment (DaughterOfMars) [reply sdk/src/wallet/account/types/balance.rs:125]: So then shouldn't this convert to the DTO and serialize with that impl?

Comment (thibault-martinez) [reply sdk/src/wallet/account/types/balance.rs:125]: Not sure what you mean but we do NOT want the DTO serialised form here

Comment (DaughterOfMars) [reply sdk/src/wallet/account/types/balance.rs:125]: Oh, I see.

File: sdk/src/types/block/output/feature/metadata.rs (hunk 2 above)
Comment (thibault-martinez) [reply sdk/src/types/block/output/feature/metadata.rs:38]: Can't, needs to be fallible because of the size

**Review by DaughterOfMars — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Pull request merged**